pkg/cidr/cidrset.go

```diff
@@ -0,0 +1,89 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cidr
+
+import (
+	"fmt"
+	"net"
+	"strings"
+)
+
+// Set contains a list of CIDRs
+type Set struct {
+	cidrs []net.IPNet
+}
+
+// NewSet builds a cidr.Set, parsing the strings
+func NewSet(cidrs []string) (Set, error) {
+	var ipNets []net.IPNet
+	for _, s := range cidrs {
+		_, ipNet, err := net.ParseCIDR(s)
+		if err != nil {
+			return Set{}, fmt.Errorf("CIDR %q not valid: %w", s, err)
+		}
+		ipNets = append(ipNets, *ipNet)
+	}
+	return Set{cidrs: ipNets}, nil
+}
+
+// WhereIPV4 returns a CIDRSet containing only the IPv4 CIDRs
+func (s Set) WhereIPV4() Set {
+	var matching []net.IPNet
+	for i := range s.cidrs {
+		if s.cidrs[i].IP.To4() != nil {
```

Comment:
Probably needs to exclude IPv4-mapped IPv6 addresses. See `validation.validateCIDR()`.

Reply:
I didn't understand the reference to validateCIDR - was it checking that ip == cidr.IP?

IIUC these IPv4-mapped IPv6 addresses are not publicly routable, and are probably going to be special cases for most networking operations on clouds.  We can add them if people start using them, but I think right now they will be an error, which I think is the correct thing to do until we/I understand better what we should do here.

I could change validateCIDR (and friends) to reject them in the validation phase though?

Reply:
`validateCIDR` uses `ip.To4() != nil && !strings.Contains(cidr, ":")` to exclude IPv4-mapped IPv6 addresses from the IPv4 code path.

Reply:
Gotcha - makes sense - thanks!  Added!